### ISSUE TITLE
Strategy pattern + abstract class

### ISSUE THREAD

**Issue by S1SYPHOS** [rosell-dk/webp-convert#22]
.. does this sound good to you? Reason going for an abstract `Converter` base class is that it allows to contain basic implementation (common tasks shared by all converters), whereas interfaces just lay out the blueprint for what each converter class should implement ..

**Comment by rosell-dk** [comment]
We could also simply create a helper class, along with a `ConverterBase` for the individual converters.

I'd like the individual converters to be able to create writable folder, test if target file exists, etc. But I also think that this stuff should be done by WebPConvert before starting a batch. So it seems both `WebPConvert` and `ConverterBase` needs `isValidTarget `, `isAllowedExtension` and `createWritableFolder` at their disposal. Hence

I'm not sure how you think of applying the Strategy pattern. I haven't used it before. Did a quick reading. It seems a bit overkill.

**Comment by S1SYPHOS** [comment]
Well, it's not exactly overkill, since `WebPConvert` acts as Context already, and with different converters being strategies, all that's left is .. an abstract base class applying mentioned functions upon being called and holding functions commonly needed, eg `escapeFilename` or whatever :lollipop:

**Comment by rosell-dk** [comment]
I don't see the `Strategy` pattern applied there... ?
Anyway, your solution doesn't allow WebPConfig to also do the checks.

I have just commited a solution that uses a helper class, which is called from both `WebPConvert` and `ConvertBase`.

**Comment by rosell-dk** [comment]
I can see that your solution also will work fine, though. The reason I wanted WebPConvert to also do the tests was that it can then exit early, if something general is wrong. There is no need to ask all converters to convert a file that does not exist. But actually, the kinds of Exceptions thrown (`TargetNotFoundException` etc) will bubble all the way up anyway. So there is no need for WebPConvert to do the checks anyway, it seems.
I will remove the helper and move the code to `ConverterBase`

**Comment by rosell-dk** [comment]
For some reason, I find it ok to have `createWritableFolder` etc public, when it is a helper class, but I don't like them to be public, when they are part of `ConverterBase`.

**Comment by rosell-dk** [comment]
I have removed checks in WebPConvert. They bubble up (#6a17deb..04ddb45)

**Comment by S1SYPHOS** [comment]
I did it like that because you wanted each of the converters to check again (I guess because you wanted to execute only one of them when asked to, but it doesn’t make sense.

The three general tasks should remain with WebPConvert, then?!

**Comment by S1SYPHOS** [comment]
.. because one of them fails, conversion is impossible

**Comment by rosell-dk** [comment]
Well, it is a choice between avoiding code redundancy and avoiding that the checks are performed multiple times, even though not needed.

In the solution I created, I went for the first, thinking that it did not matter too much, that the checks where performed multiple times.

Though it works, it however feels a bit unclean. I have therefore changed the code (committing soon!). I have added an option to the converters, whether they should perform the checks and create the destination folder. Default is true. This way, it is possible to run the converters directly. WebPConvert will also run the checks, before looping through converters. When calling the converters, it will instruct them not to do the checks
